### PR TITLE
Reimplement experimental load-balancing tests in scripttest

### PR DESCRIPTION
**Pull request by joamaki**
This PR adds a "k8s" script command and reimplements the k8s integration test-suite in `pkg/loadbalancer/experimental` as script tests.

StateDB is bumped to v0.3.2 to fix the column ordering bug.

### CODE REVIEW

File: pkg/loadbalancer/experimental/script_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/testutils"
+	"github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+func TestScript(t *testing.T) {
+	version.Force(testutils.DefaultVersion)
```
Comment (joamaki) [inline pkg/loadbalancer/experimental/script_test.go:38]: I had an earlier version that allowed doing `--k8s-force-version=1.23` in the `txtar` files, but it really wrecks havoc with parallel testing and the `version` package globals, so ended up removing the test I had for older k8s versions (e.g. Endpoints no EndpointSlices). For now I'm just forcing the version here. Will have to revisit the `pkg/k8s/version` global variable issue in a later PR instead.

File: pkg/loadbalancer/experimental/script_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			require.NoError(t, err, "ScriptCommands")
+			maps.Insert(cmds, maps.All(script.DefaultCmds()))
+
+			return &script.Engine{
+				Cmds: cmds,
+			}
+		}, []string{}, "testdata/*.txtar")
+}
+
+func parseArgsLine(file string) []string {
```
Comment (joamaki) [inline pkg/loadbalancer/experimental/script_test.go:99]: Wonder if this should be made into a builtin feature of scripttest and the args passed to the engine constructor. Not in this PR though.

File: pkg/loadbalancer/experimental/frontend.go [hunk 3]
```diff
@@ -89,6 +89,7 @@ func (fe *Frontend) TableHeader() []string {
 		"PortName",
 		"Backends",
 		"Status",
+		"StatusKind", // FIXME: Figure out a better way to do this.
```
Comment (joamaki) [inline pkg/loadbalancer/experimental/frontend.go]: This was added so that the tests can just check whether the frontend is marked "Done" without having to deal with the non-deterministic timestamp in `Status.String()` ("Done (10ms ago"). I'm kind of inclined to just leave it this way even though you'll then see "Done (10ms ago)   Done" in the table output. Alternative would be to break the Status into multiple columns (StatusKind, StatusSince, StatusMessage)...

File: pkg/loadbalancer/experimental/config.go [hunk 4]
```diff
@@ -11,9 +11,10 @@ import (
 )
 
 type Config struct {
-	EnableExperimentalLB bool
-	RetryBackoffMin      time.Duration `mapstructure:"lb-retry-backoff-min"`
-	RetryBackoffMax      time.Duration `mapstructure:"lb-retry-backoff-max"`
+	EnableExperimentalLB          bool          `mapstructure:"enable-experimental-lb"`
+	RetryBackoffMin               time.Duration `mapstructure:"lb-retry-backoff-min"`
+	RetryBackoffMax               time.Duration `mapstructure:"lb-retry-backoff-max"`
+	TestFaultInjectionProbability float32       `mapstructure:"lb-test-fault-injection"`
```
Comment (ovidiutirla) [inline pkg/loadbalancer/experimental/config.go]: I'm quite against adding Configs only for testing purposes

File: pkg/loadbalancer/experimental/frontend.go (hunk 3 above)
Comment (ovidiutirla) [reply pkg/loadbalancer/experimental/frontend.go]: could we just split the output by space in this case for example when checking the output? splitting the columns in StatusKind / StatusSince sounds reasonable to me tho.

File: pkg/k8s/client/cell.go [hunk 5]
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
...
+			for _, file := range args[1:] {
+				b, err := os.ReadFile(s.Path(file))
+				if err != nil {
+					// Try relative to current directory, e.g. to allow reading "testdata/foo.yaml"
+					b, err = os.ReadFile(file)
+				}
+				if err != nil {
+					return nil, fmt.Errorf("failed to read %s: %w", file, err)
+				}
+				obj, gvk, err := testutils.DecodeObjectGVK([]byte(b))
```
Comment (ovidiutirla) [inline pkg/k8s/client/cell.go]: `[]byte` is redundant here 
```suggestion
				obj, gvk, err := testutils.DecodeObjectGVK(b)
```

File: pkg/loadbalancer/experimental/benchmark/benchmark.go [hunk 6]
```diff
@@ -346,3 +352,245 @@ func ServicesAndSlices(testSize int) (svcs []*slim_corev1.Service, epSlices []*k
...
+			"loadbalancer-test",
+			"Test module",
+
+			cell.Provide(
+				func() experimental.Config {
+					return experimental.Config{
+						EnableExperimentalLB:          true,
+						RetryBackoffMin:               time.Millisecond,
+						RetryBackoffMax:               time.Millisecond,
+						TestFaultInjectionProbability: 0.0,
```
Comment (ovidiutirla) [inline pkg/loadbalancer/experimental/benchmark/benchmark.go]: should we use `failureProbability`?

File: pkg/loadbalancer/experimental/lbmaps.go [hunk 7]
```diff
@@ -60,6 +67,100 @@ func newLBMapsConfig(dcfg *option.DaemonConfig) (cfg LBMapsConfig) {
...
+					}
+					b, err := os.ReadFile(s.Path(args[1]))
+					if err != nil {
+						return nil, fmt.Errorf("read %q: %w", args[1], err)
+					}
+
+					expected := difflib.SplitLines(string(b))
+
+					var diff string
+					for range 10 {
```
Comment (ovidiutirla) [inline pkg/loadbalancer/experimental/lbmaps.go]: I'm not quite following why we need this for loop?

File: pkg/loadbalancer/experimental/lbmaps.go [hunk 8]
```diff
@@ -242,6 +343,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
 	return errors.Join(errs...)
 }
 
+// iteratemap iterates over a BPF map, yielding new keys and values. Similar to
```
Comment (ovidiutirla) [inline pkg/loadbalancer/experimental/lbmaps.go]:
```suggestion
// iterateMap iterates over a BPF map, yielding new keys and values. Similar to
```

File: pkg/loadbalancer/experimental/script_test.go (hunk 1 above)
Comment (ovidiutirla) [reply pkg/loadbalancer/experimental/script_test.go:38]: Let's create an issue and link it here maybe

**Review by ovidiutirla — APPROVED** [review]
LGTM 🚀

File: pkg/loadbalancer/experimental/config.go (hunk 4 above)
Comment (ovidiutirla) [reply pkg/loadbalancer/experimental/config.go]: Not super familiar with the feature itself but would be nice if we could separate the testing code from the production code.

**Comment by DamianSawicki** [comment]
> /cc @DamianSawicki

Jussi already mentioned this to me on Slack, but thanks for remembering to let me know @ovidiutirla! I'll try to have a look.

**Review by nebril — APPROVED** [review]

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 7 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go]: Reconciliation is async so we don’t know exactly when it’s done and hence need to try few times. Same as with ”db cmp”. I might think about adding some support to reconciler to allow some sort of syncing for tests.

File: pkg/loadbalancer/experimental/config.go (hunk 4 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/config.go]: I’ll split up this into a ”TestConfig” and make it an optional dependency and will only provide it in tests.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 7 above)
Comment (ovidiutirla) [reply pkg/loadbalancer/experimental/lbmaps.go]: Oh, I see now, we exit early if the len(diff) == 0. 

Wondering if it would make sense to extract the maxRetries and sleep as const? 
Also, I assume this pattern will be fairly common in the future, should we have some generic function for retries?

File: pkg/loadbalancer/experimental/frontend.go (hunk 3 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/frontend.go]: Split this into "Status", "Since" and "Error"

File: pkg/loadbalancer/experimental/benchmark/benchmark.go (hunk 6 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/benchmark/benchmark.go]: Yep, changed.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 7 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go]: Yeah probably later I'll think about adding some sort of "eventually" thing that wraps the commands. It's not entirely trivial so didn't look into it yet.

Moved these to consts.

File: pkg/k8s/client/cell.go [hunk 9]
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
...
+					return nil, fmt.Errorf("accessor: %w", err)
+				}
+				name := objMeta.GetName()
+				ns := objMeta.GetNamespace()
+
+				// Try to add the object to all the trackers. If one of them
+				// accepts we're good. We'll add to all since multiple trackers
+				// may accept (e.g. slim and kubernetes).
+
+				// err will get set to nil if any of the tracker methods succeed.
```
Comment (DamianSawicki) [inline pkg/k8s/client/cell.go:641]: How about something like the following, which should be self-explanatory enough not to require extra comment?
```
success := false
checkErr := func(e error) {
  success := success || e == nil
}
(...)
if !success
...
```
And if we want to collect the errors, we can add `var errs []error`,`append(errs, e)` , and `err = errors.Join(errs)`.

File: pkg/k8s/client/cell.go [hunk 10]
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
...
+				// start with a non-nil default error.
+				err = errors.New("fail")
+				checkErr := func(e error) {
+					if err != nil {
+						err = e
+					}
+				}
+				switch action {
```
Comment (DamianSawicki) [inline pkg/k8s/client/cell.go]: Nit: could we create a slice `[]testing.ObjectTracker` containing all of `fc.XXX.Tracker()`s before the switch and then just iterate over it for each case?

File: pkg/k8s/testutils/resources.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+			Ciliumv2APIResources,
+		},
+		"1.26": {
+			Corev1APIResources,
+			DiscoveryV1APIResources,
+			Ciliumv2APIResources,
+		},
+	}
+
+	Corev1APIResources = &metav1.APIResourceList{
```
Comment (DamianSawicki) [inline pkg/k8s/testutils/resources.go]: Nit: rename `Corev1APIResources -> CoreV1APIResources` in line with `DiscoveryV1APIResources`, and similarly `Ciliumv2APIResources -> CiliumV2APIResources`. Super-nit: `DiscoveryV1beta1APIResources -> DiscoveryV1Beta1APIResources`.

File: pkg/loadbalancer/experimental/lbmaps.go [hunk 12]
```diff
@@ -60,6 +107,109 @@ func newLBMapsConfig(dcfg *option.DaemonConfig) (cfg LBMapsConfig) {
...
+	lbmapsCompareRetryInterval = 50 * time.Millisecond
+)
+
+func newLBMapsCommand(m LBMaps) hive.ScriptCmdOut {
+	if f, ok := m.(*FaultyLBMaps); ok {
+		m = f.impl
+	}
+
+	var snapshot *mapSnapshots
+	errUsage := fmt.Errorf("%w: expected 'dump', 'snapshot' or 'restore'", script.ErrUsage)
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/lbmaps.go]: `cmp` seems missing in this list.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 7 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go]: Here we go: https://github.com/cilium/hive/pull/22

File: pkg/loadbalancer/experimental/script_test.go (hunk 1 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/script_test.go:38]: https://github.com/cilium/cilium/issues/35537

File: pkg/loadbalancer/experimental/lbmaps.go [hunk 13]
```diff
@@ -242,6 +392,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
 	return errors.Join(errs...)
 }
 
+// iterateMap iterates over a BPF map, yielding new keys and values. Similar to
+// ebpf.IterateWithCallback but allocates new key & value instead of reusing.
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/lbmaps.go:396]: What's the motivation? Is this lack of reusing a safeguard for the case when cb is non-blocking?

File: pkg/loadbalancer/experimental/lbmaps.go [hunk 14]
```diff
@@ -242,6 +392,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
...
+func iterateMap(m *ebpf.Map, newPair func() (any, any), cb func(any, any)) error {
+	entries := m.Iterate()
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/lbmaps.go:398]: The original `ebpf.IterateWithCallback` calls `m.lock.RLock()`. Why is it not needed here?

File: pkg/loadbalancer/experimental/reflector.go [hunk 15]
```diff
@@ -153,8 +154,6 @@ func runResourceReflector(ctx context.Context, p reflectorParams, initComplete f
 			}
 			txn := p.Writer.WriteTxn()
 			for _, ev := range buf {
-				ev.Done(nil)
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/reflector.go:156]: This change looks unrelated to the PR. What's the rationale behind it?

File: pkg/loadbalancer/experimental/script_test.go (hunk 2 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/script_test.go:99]: Done in https://github.com/cilium/hive/pull/22

File: pkg/k8s/client/cell.go (hunk 9 above)
Comment (joamaki) [reply pkg/k8s/client/cell.go:641]: We actually don't want to collect the errors as we do expect some of them to fail.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 13 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go:396]: `ebpf.IterateWithCallback` reused the same allocated values which means we can't use it for snapshotting maps.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 14 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go:398]: No need, it's safe to call `Iterate` concurrently. I don't know why `ebpf.Map` has a lock here.

File: pkg/loadbalancer/experimental/reflector.go (hunk 15 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/reflector.go:156]: Reason was that on sudden stopping some of the events were left unhandled causing the finalizer for the event to panic. Since we don't really care about leveraging the `Resource[T]` workqueue we can just mark them handled right away.

The plan for the reflector is to rewrite it to use client-go reflector directly and not use `Resource[T]` or an informer (no point having objects duplicated in a cache.Store), so this code will have a fairly short lifespan. I can move the change into its own commit if you'd like.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 14 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go:398]: Side note: this code really shouldn't use `pkg/ebpf` or `pkg/bpf`. We want to mostly get rid of them in the future and use `cilium/ebpf` directly. Might rewrite this pretty soon as there's very little in `pkg/ebpf` that it's building on.

File: pkg/loadbalancer/experimental/reflector.go (hunk 15 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/reflector.go:156]: > I can move the change into its own commit if you'd like.

That would be great, thanks!

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 14 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/lbmaps.go:398]: > I don't know why ebpf.Map has a lock here.

Perhaps the reason is to obtain a consistent state in case write happens concurrently?

File: pkg/loadbalancer/experimental/script_test.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+func TestScript(t *testing.T) {
+	// version/capabilities are unfortunately a global variable, so we're forcing it here.
+	// This makes it difficult to have different k8s version/capabilities (e.g. use Endpoints
+	// not EndpointSlice) in the tests here, which is why we're currently only testing against
+	// the default.
+	// Issue for fixing this: https://github.com/cilium/cilium/issues/35537
+	version.Force(testutils.DefaultVersion)
+
+	// pkg/k8s/endpoints.go uses this in ParseEndpointSlice*
+	option.Config.EnableK8sTerminatingEndpoint = true
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/script_test.go:41]: Don't we want to defer restoring the original value?

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 14 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go:398]: The writes don't take the lock. It's probably just for guarding the `OpenOrCreate` call, which we don't do here.

File: pkg/k8s/client/cell.go (hunk 9 above)
Comment (joamaki) [reply pkg/k8s/client/cell.go:641]: Ah actually useful to collect them as the error might be interesting, e.g. you might have 4 trackers rejecting it as unknown object and 1 rejecting it as known but malformed in some way. Changed this to collect all errors.

File: pkg/k8s/client/cell.go (hunk 10 above)
Comment (joamaki) [reply pkg/k8s/client/cell.go]: Good idea! Changed it so that the trackers are collected where the fake clients are created.. that way it's easy to extend.

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 12 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/lbmaps.go]: added.

File: pkg/k8s/testutils/resources.go (hunk 11 above)
Comment (joamaki) [reply pkg/k8s/testutils/resources.go]: Done

File: pkg/loadbalancer/experimental/reflector.go (hunk 15 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/reflector.go:156]: Done

File: pkg/k8s/client/cell.go (hunk 9 above)
Comment (joamaki) [reply pkg/k8s/client/cell.go:641]: e.g. here's how "double add" looks like now:
```
        scripttest.go:212: FAIL: testdata/hostport.txtar:13: k8s add pod.yaml: none of the trackers of FakeClientset accepted *v1.Pod
            mcs: no kind is registered for the type v1.Pod in scheme "sigs.k8s.io/mcs-api/pkg/client/clientset/versioned/fake/register.go:30"
            kubernetes: no kind is registered for the type v1.Pod in scheme "k8s.io/client-go/kubernetes/fake/register.go:82"
            apiexit: no kind is registered for the type v1.Pod in scheme "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake/register.go:31"
            slim: pods "my-app-85f46c4bd9-nnk25" already exists
            cilium: no kind is registered for the type v1.Pod in scheme "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/fake/register.go:18"
```

earlier you probably would've just seen "no kind is registered" error and not the "already exists" that you wanted.

File: pkg/loadbalancer/experimental/testdata/nodeport.txtar [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    protocol: TCP
+    targetPort: 80
+  selector:
+    name: echo
+  sessionAffinity: None
+  type: NodePort
+status:
+  loadBalancer: {}
+
+-- endpointslice.yaml --
```
Comment (DamianSawicki) [inline pkg/loadbalancer/experimental/testdata/nodeport.txtar:203]: I see the old yamls contained 9 endpoints each and the new ones contain 4 endpoints each. Is this change intentional? Are there any other changes? From the commit description, this seems to be a simple copy-paste commit, so I'm not sure if anyone is reviewing the details.

File: pkg/k8s/client/cell.go (hunk 9 above)
Comment (DamianSawicki) [reply pkg/k8s/client/cell.go:641]: Thanks! My main point was to use the `success` bool to avoid the need for comments (the current version has even two), but error collection and tracker iteration are also very nice changes (and your solution without bool has the benefit of avoiding the `bool+error` redundancy from my suggestion).

File: pkg/loadbalancer/experimental/lbmaps.go (hunk 14 above)
Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/lbmaps.go:398]: I see, thanks!

**Review by DamianSawicki — COMMENTED** [review]
Thanks for all the changes and explanations!

File: pkg/loadbalancer/experimental/testdata/nodeport.txtar (hunk 17 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/testdata/nodeport.txtar:203]: I decided to simplify it. There wasn't really much point in having so many.

File: pkg/loadbalancer/experimental/script_test.go (hunk 16 above)
Comment (joamaki) [reply pkg/loadbalancer/experimental/script_test.go:41]: Hmm restore it for whom? If some other test in this package depends on the value of this, then it should set it itself. Note that `option.Config` isn't actually initialized/populated, so it's in completely undefined state anyway...
Ugh I really want to get rid of these damn globals.

Comment (DamianSawicki) [reply pkg/loadbalancer/experimental/script_test.go:41]: Well, asking since not really sure, I was wondering that either now or in the future there could be some other test running afterwards? If not, let's leave it as it is for brevity.

**Review by tklauser — APPROVED** [review]

File: pkg/k8s/client/cell.go [hunk 18]
```diff
@@ -517,10 +549,30 @@ func NewFakeClientset() (*FakeClientset, Clientset) {
...
+	fd := client.KubernetesFakeClientset.Discovery().(*fakediscovery.FakeDiscovery)
+	fd.FakedServerVersion = toVersionInfo(version)
+
 	client.clientsetGetters = clientsetGetters{&client}
 	return &client, &client
 }
 
+func toVersionInfo(rawVersion string) *versionapi.Info {
+	parts := strings.Split(rawVersion, ".")
+	return &versionapi.Info{Major: parts[0], Minor: parts[1]}
```
Comment (dylandreimerink) [inline pkg/k8s/client/cell.go:573]: Is there a chance this is being called when running Cilium in a user environment? If so we should add guard here to prevent a panic on wrong input.

File: pkg/loadbalancer/experimental/benchmark/benchmark.go [hunk 19]
```diff
@@ -346,3 +352,243 @@ func ServicesAndSlices(testSize int) (svcs []*slim_corev1.Service, epSlices []*k
...
+	}
+)
+
+func testHive(maps experimental.LBMaps,
+	services chan resource.Event[*slim_corev1.Service],
+	pods chan resource.Event[*slim_corev1.Pod],
+	endpoints chan resource.Event[*k8s.Endpoints],
+	writer **experimental.Writer,
+	db **statedb.DB,
+	bo **experimental.BPFOps,
```
Comment (dylandreimerink) [inline pkg/loadbalancer/experimental/benchmark/benchmark.go:503]: I understand this C style of doing things and why in this case. But its very un-golang like, I feel this deserves some inline comments for the uninitiated.

**Review by dylandreimerink — APPROVED** [review]
Got two nits, looks good to me otherwise

File: pkg/k8s/client/cell.go (hunk 18 above)
Comment (joamaki) [reply pkg/k8s/client/cell.go:573]: Only used by the fake client currently.
